### DISCUSSION TITLE
header-includes: metadata and defaults file inconsistencies

### DISCUSSION THREAD

**Discussion by periodicpoint** [jgm/pandoc#10416] [Q&A]
### Context
I am transpiling form Markdown to PDF via XeLaTeX.

#### Pandoc version
pandoc 3.1.11.1
Features: +server +lua
Scripting engine: Lua 5.4

#### OS
Arch Linux
TeXLive: texlive-full Version: 1:2024.20240331-1

### Goal
I want to add additional LaTeX packages ad hoc without changing the existing `template.latex`.

#### Case 1 (expected to work)
If I include the extra LaTeX package via `header-includes:` in my metadata file and link it from my defaults file, everything works as expected:

File: metadata.yaml
```
header-includes:
 - '\usepackage{libertinus}'
```

#### Case 2 (expected to not work)
If I include the extra LaTeX package via `header-includes:` in my defaults file it is **not** working, which is at least somewhat **expected** (although the docs are not clear in this case):

File: defaults.yaml
```
header-includes:
 - '\usepackage{libertinus}'
```

#### Case 3 (expected to work)
If I add the extra LaTeX package via `include-in-header:` in my defaults file it is **not** working, which is totally **unexpected**.

```
include-in-header:
  - 'inc.tex' # Containing the line \usepackage{libertinus}
```

This is unexpected in particular because the docs say: 

> header-includes
>     contents specified by [-H/--include-in-header](https://pandoc.org/MANUAL.html#option--include-in-header) (may have multiple values)
(https://pandoc.org/MANUAL.html#variables-set-automatically)

#### Doc A
> Raw content to include in the document’s header may be specified using header-includes; however, it is important to mark up this content as raw code for a particular output format, using the [raw_attribute extension](https://pandoc.org/MANUAL.html#extension-raw_attribute), or it will be interpreted as Markdown. For example:
> ``` 
> header-includes:
> - |
>   ```{=latex}
>   \let\oldsection\section
>   \renewcommand{\section}[1]{\clearpage\oldsection{#1}}
>   ```
> ```

(https://pandoc.org/MANUAL.html#extension-yaml_metadata_block) 

#### Doc B
> To override or extend some [CSS](https://developer.mozilla.org/en-US/docs/Learn/CSS) for just one document, include for example:
> 
> ```
> ---
> header-includes: |
>   <style>
>   blockquote {
>     font-style: italic;
>   }
>   tr.even {
>     background-color: #f0f0f0;
>   }
>   td, th {
>     padding: 0.5em 2em 0.5em 0.5em;
>   }
>   tbody {
>     border-bottom: none;
>   }
>   </style>
> ---
> ```
(https://pandoc.org/MANUAL.html#variables-for-html)

#### Doc C
> make \paragraph and \subparagraph (fourth- and fifth-level headings, or fifth- and sixth-level with book classes) free-standing rather than run-in; requires further formatting to distinguish from \subsubsection (third- or fourth-level headings). Instead of using this option, [KOMA-Script](https://ctan.org/pkg/koma-script) can adjust headings more extensively:
> 
> ```
>     ---
>     documentclass: scrartcl
>     header-includes: |
>       \RedeclareSectionCommand[
>         beforeskip=-10pt plus -2pt minus -1pt,
>         afterskip=1sp plus -1sp minus 1sp,
>         font=\normalfont\itshape]{paragraph}
>       \RedeclareSectionCommand[
>         beforeskip=-10pt plus -2pt minus -1pt,
>         afterskip=1sp plus -1sp minus 1sp,
>         font=\normalfont\scshape,
>         indent=0pt]{subparagraph}
>     ...
> ```
(https://pandoc.org/MANUAL.html#layout)

#### Doc D
```
include-before-body:
  - inc.tex
```
(https://pandoc.org/MANUAL.html#general-writer-options-1)

### Questions

1. In doc case A it seems that we need raw code fencing, yet in doc case B and C the manual seems to suggest that fencing is not needed. **Why is this not consistent?**
2. Doc case A-C seem to cover metadata variables that should go into metadata files. In case of D there seems to be another general option to include additional TeX content via `include-before-body:`. Yet, including extra latex packages via this variable does **not** work, because there seems to be no such variable in the` template.latex` file. (The `header-includes:` is present in the template file.) **Why is this so?**
3. Why are extra LaTeX files handled via the metadata file, while they are core components of the document and most definitely **not** metadata?
4. The docs say:

```
> header-includes
>     contents specified by -H/--include-in-header (may have multiple values) 
> include-before
>     contents specified by -B/--include-before-body (may have multiple values) 
> include-after
>     contents specified by -A/--include-after-body (may have multiple values)
```
(https://pandoc.org/MANUAL.html#variables-set-automatically)
Why then, does the latex template not simply contain the corresponding variables `header-includes`, `include-before` and `include-after` directly instead of `header-includes`, `include-before` and `include-after`?

### Potentially related question
https://github.com/jgm/pandoc/discussions/10361

**Comment by lodwicz** [comment]
Seems that nobody listens, and I fear I've just stumbled upon exactly the same issue...

**Comment by jgm** [comment]
`include-in-header` in a defaults file takes a *file path* as its content (just like the command line option -- in general, defaults file fields correlate to the same-named command line options and are basically just ways of setting default values for these). So you can't put literal content there.

Put this instead in `header-includes` inside `variables`:

```
variables:
  header-includes: |
    \include{whatever}
```

This will set the `header-includes` template variable, which is what the `--include-in-header` command-line option does with the contents of the given file.

**Reply by lodwicz** [reply]
still, I think I caught a bug. When stating in the defaults file
```
variables:
  classoption: twocolumn
  header-includes:
  - |
    ```{=latex}
    \usepackage[all]{nowidow}
    ```
```
I get in the LaTeX output
```
     56 ```{=latex}
     57 \usepackage[all]{nowidow}
     58 ```
```
with line numbers produced by the output of the piped `less -N` to debug; whereas if I type in the document metadata
```
---
  header-includes:
  - |
    ```{=latex}
    \usepackage[all]{nowidow}
    ```
...
```
I get
```
     55 \setcounter{secnumdepth}{-\maxdimen} % remove section numbering
     56 \usepackage[all]{nowidow}
     57 \ifLuaTeX
```
and my document runs without problem.

Checked with 
```
pandoc 3.9.0.2
Features: +server +lua
Scripting engine: Lua 5.4
```

**Comment by jgm** [comment]
To answer the other questions,

> In doc case A it seems that we need raw code fencing, yet in doc case B and C the manual seems to suggest that fencing is not needed. Why is this not consistent?

In A you see a special kind of fencing for passing through raw content directly to a format (see `raw_attribute` in the manual).
It's fine to use that any time you want to specify raw content in a metadata field.  Sometimes you don't need it, because the markdown reader has special heuristics for recognizing raw tex and raw HTML, but these can fail in complex cases.

> Doc case A-C seem to cover metadata variables that should go into metadata files. In case of D there seems to be another general option to include additional TeX content via include-before-body:. Yet, including extra latex packages via this variable does not work, because there seems to be no such variable in the template.latex file. (The header-includes: is present in the template file.) Why is this so?

Pay attention to the context. Case D is in the section describing defaults files. YAML in a defaults file is not the same as YAML in markdown metadata.

> Why are extra LaTeX files handled via the metadata file, while they are core components of the document and most definitely not metadata?

It works because template variables are automatically populated from the corresponding metadata fields, when not otherwise set. But conceptually, it is better not to use metadata fields for this kind of thing. You can use `--include-in-header` with a file, or you can directly set the `header-includes` variable using `variables:` in a defaults file, as noted above.

The docs say:
> header-includes
>     contents specified by -H/--include-in-header (may have multiple values) 
> include-before
>     contents specified by -B/--include-before-body (may have multiple values) 
> include-after
>     contents specified by -A/--include-after-body (may have multiple values)
(https://pandoc.org/MANUAL.html#variables-set-automatically)
> Why then, does the latex template not simply contain the corresponding variables header-includes, include-before and include-after directly instead of header-includes, include-before and include-after?

??  It does include these variables. I guess I don't understand the qusetion.

**Comment by bpj** [comment]
In my custom templates (which I vimdiff with the official templates right after I install a new release, usually done in well under 20 minutes.) I have, for example in the latex template 

``````pandoc_template
$for(include-latex-before-header)$
$include-latex-before-header$

$endfor$

$for(include-latex-after-header)$
$include-latex-after-header$

$endfor$

$for(include-latex-before-body)$
$include-latex-before-body$

$endfor$

$for(include-latex-after-body)$
$include-latex-after-body$

$endfor$
``````

These are placed respectively

-   Right after `\documentclass...`,
-   Right after `$for(header-includes)$...$endfor$`,
-   Right before `$for(include-before)$...$endfor$`
-   Right after `$for(include-after)$...$endfor$`

This has several advantages:

1.  They are wholly decoupled from the `-H`/`-B`/`-A` options.
2.  I can use *both* those options to include files
    and these to include content from metadata (or the `variables:` section of a defaults file).
3.  Each is format specific so I can have stuff for
    several formats in the same document's metadata
    without conflict.

It's still a good idea to wrap material included from
metadata in explicit raw blocks though, and you need a block scalar because the
backtick character (and the `@`) is reserved "for future use" in YAML:

``````yaml
include-latex-after-header: |
  ````{=latex}
  \let\origtexttt\texttt
  \renewcommand{\texttt}[1]{\colorbox{Beige}{\origtexttt{#1}}}
  ````
``````

I would like to use an hierarchical structure like 
`$include.latex.header.after$` but for better or worse then
an `include:` in a metadata file shadows all of one in a defaults file and one in-document shadows both.
